subsys/bluetooth/audio/has.c

```diff
@@ -0,0 +1,1330 @@
+	}
+}
+
+static struct bt_gatt_cb gatt_cb = {
+	.att_mtu_updated = att_mtu_updated,
+};
+
+static int has_init(const struct device *dev)
+{
+	uint8_t type;
+
+	ARG_UNUSED(dev);
+
+	BUILD_ASSERT(BT_HAS_CP_TX_MAX_LEN < CONFIG_BT_L2CAP_TX_MTU);
+
+	if (IS_ENABLED(CONFIG_BT_HAS_HA_TYPE_MONAURAL)) {
+		type = BT_HAS_MONAURAL_HEARING_AID;
+	} else if (IS_ENABLED(CONFIG_BT_HAS_HA_TYPE_BANDED)) {
+		type = BT_HAS_BANDED_HEARING_AID;
+	} else {
+		type = BT_HAS_BINAURAL_HEARING_AID;
+	}
+
+	/* Initialize the supported features characteristic value */
+	has_local.features = BT_HAS_FEAT_HEARING_AID_TYPE_MASK & type;
+	if (IS_ENABLED(CONFIG_BT_HAS_PRESET_SYNC_SUPPORT)) {
+		has_local.features |= BT_HAS_FEAT_BIT_PRESET_SYNC;
+	}
+	if (!IS_ENABLED(CONFIG_BT_HAS_IDENTICAL_PRESET_RECORDS)) {
+		has_local.features |= BT_HAS_FEAT_BIT_INDEPENDENT_PRESETS;
+	}
+
+	bt_gatt_cb_register(&gatt_cb);
+
+	k_work_init(&active_preset_work, active_preset_work_process);
+
+	return 0;
+}
+
+SYS_INIT(has_init, APPLICATION, CONFIG_APPLICATION_INIT_PRIORITY);
+
+#define BT_HAS_IS_LOCAL(_has) ((_has) == &has_local)
+#endif /* CONFIG_BT_HAS */
+
+int bt_has_register(struct bt_has_register_param *param, struct bt_has **out)
+{
+#if defined(CONFIG_BT_HAS)
+	bool writable_presets_support = false;
+
+	*out = &has_local;
+
+	CHECKIF(param == NULL || param->ops == NULL || param->ops->active_set == NULL)
+	{
+		return -EINVAL;
+	}
+
+	/* Check if already registered */
+	if (preset_ops) {
+		return -EALREADY;
+	}
+
+	for (int i = 0; i < ARRAY_SIZE(has_preset_list); i++) {
+		struct bt_has_preset_register_param *preset_param = NULL;
+		struct has_preset *preset = &has_preset_list[i];
+
+		/* Sort the presets in order of increasing ID */
+		for (int j = 0; j < ARRAY_SIZE(param->preset_param); j++) {
+			struct bt_has_preset_register_param *tmp = &param->preset_param[j];
+
+			if ((!preset_param || tmp->idx < preset_param->idx) &&
+			    tmp->idx > last_preset_id) {
+				preset_param = tmp;
+			}
+		}
+
+		if (!preset_param) {
+			break;
+		}
+
+		preset->id = preset_param->idx;
+		preset->properties = preset_param->properties;
+#if defined(CONFIG_BT_HAS_PRESET_NAME_DYNAMIC)
+		strncpy(preset->name, preset_param->name, ARRAY_SIZE(preset->name) - 1);
+		preset->name[ARRAY_SIZE(preset->name) - 1] = '\0';
+#else
+		preset->name = preset_param->name;
+#endif /* CONFIG_BT_HAS_PRESET_NAME_DYNAMIC */
+
+		if (IS_ENABLED(CONFIG_BT_HAS_PRESET_NAME_DYNAMIC)) {
+			/* If the server exposes at least one preset record with the Writable flag
+			 * set, then the server shall set the Writable Presets Support flag.
+			 */
+			writable_presets_support |= preset_param->properties & BT_HAS_PROP_WRITABLE;
+		}
+
+		last_preset_id = preset->id;
+	}
+
+	if (writable_presets_support) {
+		has_local.features |= BT_HAS_FEAT_BIT_WRITABLE_PRESETS;
+	}
+
+	preset_ops = param->ops;
+
+	return 0;
+#else
+	return -EOPNOTSUPP;
+#endif /* CONFIG_BT_HAS */
+}
+
+int bt_has_preset_active_get(struct bt_has *has)
```

Comment:
If this function is for the client only, we can guard it with `CONFIG_BT_HAS_CLIENT`.

Reply:
This applies to multiple functions.

Reply:
Why is this, and others, client only functions in the first place? Seems useful enough for the server as well.

Reply:
I would like to avoid situation where Bluetooth service is interface for getting Active Preset by other parts of the system.
If the Active Preset has to be exposed to the other parts of the system, then Application implements HAS should have a getter function IMO.